eskipfile/remote_test.go

```diff
@@ -189,9 +180,111 @@ func TestHTTPTimeout(t *testing.T) {
 	}
 }
 
-func createTestServer(c string, sc int) *httptest.Server {
-	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		w.WriteHeader(sc)
-		io.WriteString(w, c)
+func TestRoutesCaching(t *testing.T) {
+	count200s := atomic.Int32{}
+	count304s := atomic.Int32{}
+	expected := eskip.MustParse(fmt.Sprintf("VALID: %v;", routeBody))
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if noneMatch := r.Header.Get("If-None-Match"); noneMatch == "test-etag" {
+			t.Logf("request matches etag: %s", noneMatch)
+			w.WriteHeader(http.StatusNotModified)
+			count304s.Add(1)
+		} else {
+			w.Header().Set("ETag", "test-etag")
+			io.WriteString(w, fmt.Sprintf("VALID: %v;", routeBody))
+			count200s.Add(1)
+		}
+	}))
+	defer server.Close()
+
+	options := &RemoteWatchOptions{RemoteFile: server.URL, Threshold: 10, Verbose: true, FailOnStartup: true}
+	client, err := RemoteWatch(options) // First load done with initialization because of FailOnStartup
+
+	require.NoError(t, err)
+
+	defer client.(*remoteEskipFile).Close()
+
+	r, err := client.LoadAll()
+
+	t.Logf("uncached responses received: %d", count200s.Load())
+	assert.Equal(t, int32(1), count200s.Load())
+	t.Logf("cached responses received: %d", count304s.Load())
+	assert.Equal(t, int32(1), count304s.Load())
+
+	require.NoError(t, err)
+
+	assert.Equal(t, r, expected)
+
+}
+
+func TestRoutesCachingWrongEtag(t *testing.T) {
+	alterante := atomic.Int32{}
+	count200s := atomic.Int32{}
+	count304s := atomic.Int32{}
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if noneMatch := r.Header.Get("If-None-Match"); (alterante.Load()%2 != 0 && noneMatch == "test-etag") || (alterante.Load()%2 == 0 && noneMatch == "different-etag") {
+			w.WriteHeader(http.StatusNotModified)
+			count304s.Add(1)
+		} else {
+			if alterante.Load()%2 == 0 {
+				w.Header().Set("ETag", "different-etag")
+				io.WriteString(w, fmt.Sprintf("different: %v;", routeBody))
+			} else {
+				w.Header().Set("ETag", "test-etag")
+				io.WriteString(w, fmt.Sprintf("VALID: %v;", routeBody))
+			}
+			count200s.Add(1)
+		}
+		alterante.Add(1)
 	}))
+	defer ts.Close()
+
+	options := &RemoteWatchOptions{RemoteFile: ts.URL, Threshold: 10, Verbose: true, FailOnStartup: true}
+	client, err := RemoteWatch(options)
+	require.NoError(t, err)
+
+	defer client.(*remoteEskipFile).Close()
+
+	r, err := client.LoadAll()
+
+	require.NoError(t, err)
+
+	t.Logf("uncached responses received: %d", count200s.Load())
+	assert.Equal(t, int32(2), count200s.Load())
+	t.Logf("cached responses received: %d", count304s.Load())
+	assert.Equal(t, int32(0), count304s.Load())
+
+	expected := getExpectedRoutes(&alterante)
```

Comment:
Please inline getExpectedRoutes. It also should not depend on the alterante - at this point you know how many requests were done and should check expected result explicitly.